### PR TITLE
[DI] Failing test case for complex near-circular situation + lazy

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30091
| License       | MIT
| Doc PR        | not needed

Just the failing test case. But hopefully, this can get things moving :)

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/containers/container_almost_circular.php [hunk 1]
```diff
@@ -102,6 +103,34 @@
...
+$container->register('connection3', 'stdClass')
+    ->setProperty('listener', [new Reference('listener3')]);
+
+$container->register('listener3', 'stdClass')
+    ->setPublic(true) // just for assertions
+    ->setProperty('manager', new Reference('manager3'));
+
+// doctrine-like event system with small differences
+
+$container->register('manager4', 'stdClass')
```
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/containers/container_almost_circular.php]: This is a second case that's actually unrelated, but I stumbled upon it. For some reason, if `connection4` is NOT public, you get a nesting exception.

File: src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php [hunk 2]
```diff
@@ -1396,6 +1396,14 @@ public function testAlmostCircular($visibility)
         $this->assertEquals((object) ['bar6' => (object) []], $foo6);
 
         $this->assertInstanceOf(\stdClass::class, $container->get('root'));
+
+        $manager3 = $container->get('manager3');
+        $listener3 = $container->get('listener3');
+        $this->assertSame($manager3, $listener3->manager, 'Both should identically be the manaer3 service');
```
Comment (Lctrs) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerBuilderTest.php]:
```suggestion
        $this->assertSame($manager3, $listener3->manager, 'Both should identically be the manager3 service');
```

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 3]
```diff
@@ -906,6 +906,14 @@ public function testAlmostCircular($visibility)
         $this->assertEquals((object) ['bar6' => (object) []], $foo6);
 
         $this->assertInstanceOf(\stdClass::class, $container->get('root'));
+
+        $manager3 = $container->get('manager3');
+        $listener3 = $container->get('listener3');
+        $this->assertSame($manager3, $listener3->manager, 'Both should identically be the manaer3 service');
```
Comment (Lctrs) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]:
```suggestion
        $this->assertSame($manager3, $listener3->manager, 'Both should identically be the manager3 service');
```

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/containers/container_almost_circular.php (hunk 1 above)
Comment (Lctrs) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/containers/container_almost_circular.php]: Could be worth to add this explanation as a comment code ?

**Comment by nicolas-grekas** [comment]
Thank you @weaverryan 
I've added you test cases to #30096, they were perfect.
I've also pushed here the patch that should be applied to branch 4.3, when merging up.

**Pull request closed without merging**